### PR TITLE
Use Redis-based sessions if Redis is available

### PR DESCRIPTION
**Pull request by Crell**
Enable Redis by default for Platform.sh usage.

* Turn Redis on, ever and always.  There's no compelling reason for someone to not use it on Platform.sh.
* Define a `rediscache` service and endpoint, which will get used for cache if available.
* Optionally define a `redissession` service and endpoint, commented out.  If a particular user desires to use a separate Redis instance they can by uncommenting a few lines.  If not, `rediscache` gets used for sessions as well.
* Remove the Memcache glue code, because we want people to use Redis instead.
* If for some reason Redis is broken, the system will still fall back to disk-based cache and sessions.  But really, try not to do that. :smile: 

This branch once accepted should be forward-ported to 2.x.

I spoke with a member of our support team, and he said there's nothing wrong with having 2 Redis instances (a number of our customers do for various reasons), but we don't know enough about eZ's usage profile to say if it's useful or recommended.  So split the difference for now with an optional toggle.  If we find over time that it really does help in most cases then we can default the separate session store to on very easily.

### CODE REVIEW

**Comment by crevillo** [comment]
Is not this addition quite similar to which we already have in https://github.com/Crell/ezplatform/blob/989bd9165816f83fdd0e8dfaf768ef77c91e7b36/app/config/env/platformsh.php#L36?
Or in other those, why doing another look and not move your ini_set commands to the part when params are set?

**Comment by andrerom** [comment]
@Crell It wouldn't it be better if we autodetect `sessionstorage` as proposed in #234? At least from what I was told from others it was advisable to use `rediscache:redis` for cache, an `redisdata:redis` for session to make sure sessions are not randomly deleted, while making sure cache in placed in fast in-memory configured redis where  is it _does_ delete older data when it goes over defined memory limits.

**Comment by Crell** [comment]
Based on discussion with @andrerom earlier today, here's a revised version, now built against 1.13.  I'll update the OP momentarily with a description.

File: app/config/env/platformsh.php [hunk 1]
```diff
@@ -24,28 +24,44 @@
...
-        $container->setParameter('cache_memcached_port', $endpoint['port']);
+        $container->setParameter('cache_redis_port', $endpoint['port']);
     }
-} elseif (isset($relationships['redis'])) {
-    foreach ($relationships['redis'] as $endpoint) {
+}
+
+// Use Redis-based sessions if possible. If a separate Redis instance
+// is available, use that.  If not, share a Redis instance with the
+// Cache.  (That should be safe to do except on especially high-traffic sites.)
```
[the remark this answers is not in the document]
Comment (andrerom) [reply app/config/env/platformsh.php]: see above on persistenc, the optional seperation was suggested in cases where data cache is filling up the cache and potentially start to evict sessions way before it's needed. Typically useful if you have lots of content that will fill up the main redis cache.

**Comment by andrerom** [comment]
@vidarl no, but as long as we don't use persistent instance, should not be a big problem, or? (expiry time will be set according to settings in php.ini afaik)

File: .platform.app.yaml [hunk 2]
```diff
@@ -18,7 +18,11 @@ build:
 # to the application in the PLATFORM_RELATIONSHIPS variable. The right-hand
 # side is in the form `<service name>:<endpoint name>`.
 relationships:
-    database: "mysqldb:mysql"
+    database: 'mysqldb:mysql'
+    rediscache: 'rediscache:redis'
+    # If you wish to have a separate Redis instance for sessions, uncomment
+    # this relationship and the corresponding service in .platform.app.yaml.
```
[the remark this answers is not in the document]
Comment (andrerom) [reply .platform.app.yaml]: yes

**Comment by vidarl** [comment]
@andrerom : Not a problem as long as we don't use persistance instance true

File: app/config/env/platformsh.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (vidarl) [reply app/config/env/platformsh.php]: Yeah, but I am not so sure enterprise customers is keen on treating session info as cache ( disposable ). So for instance during the announced maintenance windows at platform.sh lately due to Meltdown and Spectre where the servers are rebooted, it also means that customer looses all their session data if that is stored in redis.
Have you considered  this ?

[the remark this answers is not in the document]
Comment (andrerom) [reply app/config/env/platformsh.php]: @Crell Fair point here. Also didn't you mention cache will be wiped when deploying new branches or something along those lines?

**Comment by andrerom** [comment]
@vidarl / @Crell Are you ok with this as of aa07e74?

File: app/config/env/platformsh.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrerom) [reply app/config/env/platformsh.php]: @vidarl added in aa07e74

[the remark this answers is not in the document]
Comment (Crell) [reply app/config/env/platformsh.php]: The configuration here applies to PS only.  On PE, the `services.yaml` file is ignored and controlled from the Platform side.  We can have different defaults on PE and PS if we want.

The idea of splitting the sessions off is so that they have a separate LRU limit; that is, a hyper-active cache is not going to start logging people out.  In practice I don't expect a dedicated sessions Redis to ever fill up on its own.  On a site with only a few sessions (authenticated users), I wouldn't expect it to be much of an issue but I'm less familiar with eZ's usage patterns here.

With memory-resident Redis, data would be cleared when:

* Cloning to a new environment (the new env will start with empty Redis)
* When the Redis service restarts (which should be rare, as it doesn't need to restart when only the app container is updated).

So I expect it to be reasonably safe.  Memory-resident Redis is what the vast majority of our Redis-for-cache installs are.

That said, there's no reason we can't default eZ to persistent Redis.  The only significant drawback is that we'd need to allocate disk space for it, and disk space starts at only 5 GB on PS.  That means we'd need to decide what size to give it, and ensure that the Redis, MySQL, and file system capacity still fits under 5 GB.  (And maybe leave room for someone to add Solr?)  I don't know what an appropriate size is for either Redis pool in that case.

In short: I'm flexible here.  I think for PS one-with-optional memory Redis is fine, but if you'd rather go with 2 persistent out of the box I am fine with that, too.  Just let me know what appropriate disk sizes should be for the various services and app disk.

PE is another animal entirely: For fail-over reasons persistent redis makes a lot more sense there, but we can discuss that separately.  As long as the service names line up the PHP code here should "just work" in either case, whatever we end up doing.

[the remark this answers is not in the document]
Comment (andrerom) [reply app/config/env/platformsh.php]: See the text updates @Crell, intention is to recommend cache instance over persistence but just mention it for those that want it. But this is only for session, using persistence for ca hint would be utterly backwards.

**Comment by andrerom** [comment]
All is in agreement up until this point: `where redissession is persistent`
Based on info form @Crell this is not common, which is why I word it here in a way that end user needs to pick one or the other depending on their preferences/requirements.

**Comment by vidarl** [comment]
@crell: Did you actually test this code?

I don't get redis cache to work by using the injection in config/env/platformsh.php.
I do get it work if I hard code the same settings in .platform.app.yml though:

```
variables:
    php:
        "session.save_handler": "redis"
        "session.save_path": "tcp://rediscache.internal:6379"
```

**Comment by Crell** [comment]
I tested it to verify that the ini settings were being set correctly.  I'm unsure how to test the application level to ensure it's reading from Redis.

It... should work as long as those ini settings are set prior to the session being initialized.  When does eZ first initialize its session?  I assumed it would be after that file is included.

**Comment by vidarl** [comment]
The problem is these [lines](https://github.com/Crell/ezplatform/blob/a609192df1b9dd039ac1cce2ffea4e7d3b65272e/app/config/env/platformsh.php#L58-L59))

app/config/env/platformsh.php is not executed on every pageload, only at compile time. This the inisetting is also change during compile time.

I'll look into a fix

**Comment by vidarl** [comment]
According to https://docs.platform.sh/configuration/services/redis.html#using-redis-as-handler-for-native-php-sessions it is fine to hardcode the redis config using php settings.

I asked #platformsh on slack and they confirmed that the config indeed is deterministic.
@Crell : I'll write a new PR where the Redis settings is injected as documented.

One downside of this approach though is that if you are using PE with wings, you'll need different settings (different .platform.app.yaml) for master, staging and production environments.

**Comment by Crell** [comment]
I'm honestly torn and may remove that section from the docs for exactly the reason you mention: It makes it impossible to use with Wings, because yo ucannot have a separate .platform.app.yaml file then.  That section was added by Sebastian before the tighter PS->PE integration was commonplace.  Since it's incompatible with PE, though, it's not an option here either way.
